Framework/Algorithms/src/AbsorptionCorrectionPaalmanPings.cpp

```diff
@@ -0,0 +1,486 @@
+    assc->setSharedX(i, m_inputWS->sharedX(i));
+    acc->setSharedX(i, m_inputWS->sharedX(i));
+    acsc->setSharedX(i, m_inputWS->sharedX(i));
+
+    if (!spectrumInfo.hasDetectors(i)) {
+      g_log.information() << "Spectrum " << i
+                          << " does not have a detector defined for it\n";
+      continue;
+    }
+    const auto &det = spectrumInfo.detector(i);
+
+    std::vector<double> sample_L2s(m_numVolumeElements);
+    std::vector<double> sample_container_L2s(m_numVolumeElements);
+    std::vector<double> container_L2s(m_containerNumVolumeElements);
+    std::vector<double> container_sample_L2s(m_containerNumVolumeElements);
+    calculateDistances(det, sample_L2s, sample_container_L2s, container_L2s,
+                       container_sample_L2s);
+
+    const auto wavelengths = m_inputWS->points(i);
+    // these need to have the minus sign applied still
+    const auto linearCoefAbs =
+        m_material.linearAbsorpCoef(wavelengths.cbegin(), wavelengths.cend());
+    const auto containerLinearCoefAbs = m_containerMaterial.linearAbsorpCoef(
+        wavelengths.cbegin(), wavelengths.cend());
+
+    // Get a reference to the Y's in the output WS for storing the factors
+    auto &assY = ass->mutableY(i);
+    auto &asscY = assc->mutableY(i);
+    auto &accY = acc->mutableY(i);
+    auto &acscY = acsc->mutableY(i);
+
+    // Loop through the bins in the current spectrum every m_xStep
+    for (int64_t j = 0; j < specSize; j = j + m_xStep) {
+      double integral = 0.0;
+      double crossIntegral = 0.0;
+
+      doIntegration(integral, crossIntegral, -linearCoefAbs[j],
+                    m_linearCoefTotScatt, m_elementVolumes, m_L1s, sample_L2s,
+                    -containerLinearCoefAbs[j], m_containerLinearCoefTotScatt,
+                    m_sample_containerL1s, sample_container_L2s, 0,
+                    m_numVolumeElements);
+      assY[j] = integral / m_sampleVolume;
+      asscY[j] =
+          crossIntegral / m_sampleVolume; // Divide by total volume of the shape
+
+      integral = 0.0;
+      crossIntegral = 0.0;
+      doIntegration(integral, crossIntegral, -containerLinearCoefAbs[j],
+                    m_containerLinearCoefTotScatt, m_containerElementVolumes,
+                    m_containerL1s, container_L2s, -linearCoefAbs[j],
+                    m_linearCoefTotScatt, m_container_sampleL1s,
+                    container_sample_L2s, 0, m_containerNumVolumeElements);
+      accY[j] =
+          integral / m_containerVolume; // Divide by total volume of the shape
+      acscY[j] = crossIntegral /
+                 m_containerVolume; // Divide by total volume of the shape
+
+      // Make certain that last point is calculated
+      if (m_xStep > 1 && j + m_xStep >= specSize && j + 1 != specSize) {
+        j = specSize - m_xStep - 1;
+      }
+    }
+
+    // Interpolate linearly between points separated by m_xStep,
+    // last point required
+    if (m_xStep > 1) {
```

Comment:
Am I correct in understanding that the calculation is always done in the original binning then interpolated to the requested number of wavelength points if supplied? Would it be better to setup the wavelength bins as requested instead?

Reply:
So I just [left this in from AbsorptionCorrection](https://github.com/mantidproject/mantid/blob/a1e9dc356f470c7240feda84edb72b6f14a2bbde/Framework/Algorithms/src/AbsorptionCorrection.cpp#L281) because I thought it might be useful and I didn't even test it. My understanding is you can specify fewer point using `NumberOfWavelengthPoints` to calculate then interpolate the result, by default it calculates it for every wavelength bin.